<launch>
  <arg name="serial_no"           default=""/>
  <arg name="usb_port_id"         default=""/>
  <arg name="device_type"         default="d435i"/>
  <arg name="json_file_path"      default="$(find neptune_ros)/config/d435i_emitter_disabled.json"/>
  <arg name="camera"              default="d435i"/>
  <arg name="tf_prefix"           default="$(arg camera)"/>
  <arg name="external_manager"    default="false"/>
  <arg name="manager"             default="realsense2_camera_manager"/>
  <arg name="output"              default="screen"/>
  <arg name="respawn"             default="true"/>

  <arg name="depth_width"         default="640"/>
  <arg name="depth_height"        default="480"/>
  <arg name="enable_depth"        default="true"/>

  <arg name="infra_width"         default="640"/>
  <arg name="infra_height"        default="480"/>
  <arg name="enable_infra"        default="true"/>
  <arg name="enable_infra1"       default="true"/>
  <arg name="enable_infra2"       default="true"/>
  <arg name="infra_rgb"           default="false"/>

  <arg name="color_width"         default="640"/>
  <arg name="color_height"        default="480"/>
  <arg name="enable_color"        default="true"/>

  <arg name="depth_fps"           default="30"/>
  <arg name="infra_fps"           default="30"/>
  <arg name="color_fps"           default="30"/>
  <arg name="gyro_fps"            default="200"/>
  <arg name="accel_fps"           default="250"/>
  <arg name="enable_gyro"         default="true"/>
  <arg name="enable_accel"        default="true"/>

  <arg name="enable_pointcloud"         default="false"/>
  <arg name="pointcloud_texture_stream" default="RS2_STREAM_COLOR"/>
  <arg name="pointcloud_texture_index"  default="0"/>
  <arg name="allow_no_texture_points"   default="false"/>
  <arg name="ordered_pc"                default="false"/>

  <arg name="enable_sync"               default="true"/>
  <arg name="align_depth"               default="true"/>

  <arg name="publish_tf"                default="true"/>
  <arg name="tf_publish_rate"           default="0"/>

  <arg name="filters"                   default=""/>
  <arg name="clip_distance"             default="-2"/>
  <arg name="linear_accel_cov"          default="0.01"/>
  <arg name="initial_reset"             default="false"/>
  <arg name="unite_imu_method"          default="linear_interpolation"/>

  <arg name="stereo_module/exposure/1"  default="7500"/>
  <arg name="stereo_module/gain/1"      default="16"/>
  <arg name="stereo_module/exposure/2"  default="1"/>
  <arg name="stereo_module/gain/2"      default="16"/>

  <!-- Custom -->>
  <!-- <arg name="enable_image_view"         default="false"/> -->
  <arg name="enable_imu_filter"         default="true"/>
  <arg name="enable_throttle"           default="true"/>
  <arg name="stereo_sync"               default="true"/>

  <group ns="$(arg camera)">

    <include file="$(find realsense2_camera)/launch/includes/nodelet.launch.xml">
      <arg name="tf_prefix"                value="$(arg tf_prefix)"/>
      <arg name="external_manager"         value="$(arg external_manager)"/>
      <arg name="manager"                  value="$(arg manager)"/>
      <arg name="output"                   value="$(arg output)"/>
      <arg name="respawn"                  value="$(arg respawn)"/>
      <arg name="serial_no"                value="$(arg serial_no)"/>
      <arg name="usb_port_id"              value="$(arg usb_port_id)"/>
      <arg name="device_type"              value="$(arg device_type)"/>
      <arg name="json_file_path"           value="$(arg json_file_path)"/>

      <arg name="enable_pointcloud"        value="$(arg enable_pointcloud)"/>
      <arg name="pointcloud_texture_stream" value="$(arg pointcloud_texture_stream)"/>
      <arg name="pointcloud_texture_index"  value="$(arg pointcloud_texture_index)"/>
      <arg name="enable_sync"              value="$(arg enable_sync)"/>
      <arg name="align_depth"              value="$(arg align_depth)"/>

      <arg name="depth_width"              value="$(arg depth_width)"/>
      <arg name="depth_height"             value="$(arg depth_height)"/>
      <arg name="enable_depth"             value="$(arg enable_depth)"/>

      <arg name="color_width"              value="$(arg color_width)"/>
      <arg name="color_height"             value="$(arg color_height)"/>
      <arg name="enable_color"             value="$(arg enable_color)"/>

      <arg name="infra_width"              value="$(arg infra_width)"/>
      <arg name="infra_height"             value="$(arg infra_height)"/>
      <arg name="enable_infra"             value="$(arg enable_infra)"/>
      <arg name="enable_infra1"            value="$(arg enable_infra1)"/>
      <arg name="enable_infra2"            value="$(arg enable_infra2)"/>
      <arg name="infra_rgb"                value="$(arg infra_rgb)"/>

      <arg name="depth_fps"                value="$(arg depth_fps)"/>
      <arg name="infra_fps"                value="$(arg infra_fps)"/>
      <arg name="color_fps"                value="$(arg color_fps)"/>
      <arg name="gyro_fps"                 value="$(arg gyro_fps)"/>
      <arg name="accel_fps"                value="$(arg accel_fps)"/>
      <arg name="enable_gyro"              value="$(arg enable_gyro)"/>
      <arg name="enable_accel"             value="$(arg enable_accel)"/>

      <arg name="publish_tf"               value="$(arg publish_tf)"/>
      <arg name="tf_publish_rate"          value="$(arg tf_publish_rate)"/>

      <arg name="filters"                  value="$(arg filters)"/>
      <arg name="clip_distance"            value="$(arg clip_distance)"/>
      <arg name="linear_accel_cov"         value="$(arg linear_accel_cov)"/>
      <arg name="initial_reset"            value="$(arg initial_reset)"/>
      <arg name="unite_imu_method"         value="$(arg unite_imu_method)"/>
      <arg name="stereo_module/exposure/1" value="$(arg stereo_module/exposure/1)"/>
      <arg name="stereo_module/gain/1"     value="$(arg stereo_module/gain/1)"/>
      <arg name="stereo_module/exposure/2" value="$(arg stereo_module/exposure/2)"/>
      <arg name="stereo_module/gain/2"     value="$(arg stereo_module/gain/2)"/>

      <arg name="allow_no_texture_points"  value="$(arg allow_no_texture_points)"/>
      <arg name="ordered_pc"               value="$(arg ordered_pc)"/>
      
    </include>

    <!-- Data throttle and sync -->
    <!-- <group if="$(eval arg('enable_throttle') == 'true')"> -->
    <node if="$(arg stereo_sync)" pkg="nodelet" type="nodelet" name="stereo_sync" args="load rtabmap_ros/stereo_sync /d435i/realsense2_camera_manager" output="screen">
      <param name="compressed_rate"  type="double" value="5.0"/>
      <param name="approx_sync"      type="bool"   value="false"/>

      <remap from="left/image_rect"   to="infra1/image_rect_raw"/>
      <remap from="right/image_rect"  to="infra2/image_rect_raw"/>
      <remap from="left/camera_info"  to="infra1/camera_info"/>
      <remap from="right/camera_info" to="infra2/camera_info"/>

      <remap from="rgbd_image"        to="rgbd_image"/>
    </node>

    <node unless="$(arg stereo_sync)" pkg="nodelet" type="nodelet" name="stereo_throttle" args="load rtabmap_ros/stereo_throttle /d435i/realsense2_camera_manager" output="screen">
      <param name="rate"         type="double" value="5.0"/>
      <param name="approx_sync"  type="bool"   value="false"/>

      <remap from="left/image"        to="infra1/image_rect_raw"/>
      <remap from="right/image"       to="infra2/image_rect_raw"/>
      <remap from="left/camera_info"  to="infra1/camera_info"/>
      <remap from="right/camera_info" to="infra2/camera_info"/>

      <remap from="infra1/image_rect_raw_throttle"  to="throttled/infra1/image_rect_raw"/>
      <remap from="infra2/image_rect_raw_throttle"  to="throttled/infra2/image_rect_raw"/>
      <remap from="infra1/camera_info_throttle"     to="throttled/infra1/camera_info"/>
      <remap from="infra2/camera_info_throttle"     to="throttled/infra2/camera_info"/>
    </node>
    <!-- </group> -->
  </group>

  
  <!-- imu_filter -->
    <node pkg="imu_filter_madgwick" type="imu_filter_node" name="imu_filter_node" output="screen" 
      if="$(arg enable_imu_filter)">
      <param name="~use_mag"      value='false'/> 
      <param name="~publish_tf"   value='false'/> 
      <param name="~world_frame"  value="enu"/> 
      <remap from="/imu/data_raw" to="/$(arg camera)/imu"/> 
      <remap from="/imu/data"     to="/rtabmap/imu"/> 
    </node>
  
  <!--Visualize-->
  <!-- <node pkg="image_view" type="image_view" name="visualize" output="screen"
    if="$(arg enable_image_view)">
    <remap from="image" to="/d435i/color/image_raw"/>
    <param name="~image_transport" value="raw"/> 
  </node> -->
</launch>
